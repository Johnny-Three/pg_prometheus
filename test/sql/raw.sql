\set ECHO ALL
\set ON_ERROR_STOP 1
SET TIME ZONE 'UTC';

DROP TABLE IF EXISTS metrics CASCADE;

CREATE TABLE metrics (sample prom_sample);
CREATE INDEX metrics_time_idx ON metrics (prom_time(sample));
CREATE INDEX metrics_labels_idx ON metrics USING GIN (prom_labels(sample));

\d metrics

INSERT INTO metrics VALUES ('cpu_usage{service="nginx",host="machine1"} 34.6 1494595898000'),
                           ('cpu_usage{service="nginx",host="machine2"} 10.3 1494595899000'),
                           ('cpu_usage{service="nginx",host="machine1"} 30.2 1494595928000');

SELECT * FROM metrics;

SELECT prom_time(sample), prom_value(sample), prom_labels(sample) FROM metrics;


SELECT prom_time(sample), prom_value(sample) FROM metrics 
WHERE prom_time(sample) >  'Fri May 12 13:31:00 2017' AND
      prom_name(sample) = 'cpu_usage' AND
      prom_labels(sample) @> '{"service": "nginx", "host": "machine1"}';

EXPLAIN ANALYZE SELECT prom_time(sample), prom_value(sample) FROM metrics 
WHERE prom_time(sample) >  'Fri May 12 13:31:00 2017' AND
      prom_name(sample) = 'cpu_usage' AND
      prom_labels(sample) @> '{"service": "nginx", "host": "machine1"}';

-- Cleanup
DROP TABLE metrics CASCADE;
