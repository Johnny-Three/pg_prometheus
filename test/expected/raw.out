\set ECHO ALL
\set ON_ERROR_STOP 1
SET TIME ZONE 'UTC';
DROP TABLE IF EXISTS metrics CASCADE;
NOTICE:  table "metrics" does not exist, skipping
CREATE TABLE metrics (sample prom_sample);
CREATE INDEX metrics_time_idx ON metrics (prom_time(sample));
CREATE INDEX metrics_labels_idx ON metrics USING GIN (prom_labels(sample));
\d metrics
      Table "public.metrics"
 Column |    Type     | Modifiers 
--------+-------------+-----------
 sample | prom_sample | 
Indexes:
    "metrics_labels_idx" gin (prom_labels(sample))
    "metrics_time_idx" btree (prom_time(sample))

INSERT INTO metrics VALUES ('cpu_usage{service="nginx",host="machine1"} 34.6 1494595898000'),
                           ('cpu_usage{service="nginx",host="machine2"} 10.3 1494595899000'),
                           ('cpu_usage{service="nginx",host="machine1"} 30.2 1494595928000'),
                           -- Test with no labels
                           ('cpu_usage 30.2 1494595928000');
SELECT * FROM metrics;
                                sample                                 
-----------------------------------------------------------------------
 cpu_usage{service="nginx",host="machine1"} 34.600000 1494595898000000
 cpu_usage{service="nginx",host="machine2"} 10.300000 1494595899000000
 cpu_usage{service="nginx",host="machine1"} 30.200000 1494595928000000
 cpu_usage 30.200000 1494595928000000
(4 rows)

SELECT prom_time(sample), prom_value(sample), prom_labels(sample) FROM metrics;
          prom_time           | prom_value |               prom_labels                
------------------------------+------------+------------------------------------------
 Fri May 12 13:31:38 2017 UTC |       34.6 | {"host": "machine1", "service": "nginx"}
 Fri May 12 13:31:39 2017 UTC |       10.3 | {"host": "machine2", "service": "nginx"}
 Fri May 12 13:32:08 2017 UTC |       30.2 | {"host": "machine1", "service": "nginx"}
 Fri May 12 13:32:08 2017 UTC |       30.2 | {}
(4 rows)

SELECT prom_time(sample), prom_value(sample) FROM metrics
WHERE prom_time(sample) >  'Fri May 12 13:31:00 2017' AND
      prom_name(sample) = 'cpu_usage' AND
      prom_labels(sample) @> '{"service": "nginx", "host": "machine1"}';
          prom_time           | prom_value 
------------------------------+------------
 Fri May 12 13:31:38 2017 UTC |       34.6
 Fri May 12 13:32:08 2017 UTC |       30.2
(2 rows)

EXPLAIN SELECT prom_time(sample), prom_value(sample) FROM metrics
WHERE prom_time(sample) >  'Fri May 12 13:31:00 2017' AND
      prom_name(sample) = 'cpu_usage' AND
      prom_labels(sample) @> '{"service": "nginx", "host": "machine1"}';
                                                               QUERY PLAN                                                               
----------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on metrics  (cost=24.01..28.04 rows=1 width=16)
   Recheck Cond: (prom_labels(sample) @> '{"host": "machine1", "service": "nginx"}'::jsonb)
   Filter: ((prom_time(sample) > 'Fri May 12 13:31:00 2017 UTC'::timestamp with time zone) AND (prom_name(sample) = 'cpu_usage'::text))
   ->  Bitmap Index Scan on metrics_labels_idx  (cost=0.00..24.01 rows=1 width=0)
         Index Cond: (prom_labels(sample) @> '{"host": "machine1", "service": "nginx"}'::jsonb)
(5 rows)

-- Insert sample with no timestamp
INSERT INTO metrics VALUES ('cpu_usage{service="nginx",host="machine3"} 30.2');
-- Cannot show time since different every test run
SELECT prom_value(sample), prom_labels(sample) FROM metrics;
 prom_value |               prom_labels                
------------+------------------------------------------
       34.6 | {"host": "machine1", "service": "nginx"}
       10.3 | {"host": "machine2", "service": "nginx"}
       30.2 | {"host": "machine1", "service": "nginx"}
       30.2 | {}
       30.2 | {"host": "machine3", "service": "nginx"}
(5 rows)

-- Cleanup
DROP TABLE metrics CASCADE;
